File: backend/routes/health_data.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import db
from models.user import User
from models.health_data import HealthData
from datetime import datetime, timedelta
import pandas as pd

health_data_bp = Blueprint('health_data', __name__)

@health_data_bp.route('/', methods=['GET'])
@jwt_required()
def get_health_data():
    """Get user health data with optional filtering"""
    current_user_id = get_jwt_identity()
    user = User.query.filter_by(public_id=current_user_id).first()
    
    if not user:
        return jsonify({'message': 'User not found'}), 404
    
    # Parse query parameters
    start_date = request.args.get('start_date')
    end_date = request.args.get('end_date')
    metric = request.args.get('metric')
    limit = request.args.get('limit', default=30, type=int)
    
    # Base query
    query = HealthData.query.filter_by(user_id=user.id)
    
    # Apply filters
    if start_date:
        try:
            start_date = datetime.fromisoformat(start_date.replace('Z', '+00:00'))
            query = query.filter(HealthData.date >= start_date)
        except ValueError:
            return jsonify({'message': 'Invalid start_date format. Use ISO format.'}), 400
    
    if end_date:
        try:
            end_date = datetime.fromisoformat(end_date.replace('Z', '+00:00'))
            query = query.filter(HealthData.date <= end_date)
        except ValueError:
            return jsonify({'message': 'Invalid end_date format. Use ISO format.'}), 400
    
    # Order by date (newest first) and apply limit
    health_data = query.order_by(HealthData.date.desc()).limit(limit).all()
    
    # Convert to dictionaries
    result = [data.to_dict() for data in health_data]
    
    return jsonify(result), 200

@health_data_bp.route('/', methods=['POST'])
@jwt_required()
def add_health_data():
    """Add new health data entry"""
    current_user_id = get_jwt_identity()
    user = User.query.filter_by(public_id=current_user_id).first()
    
    if not user:
        return jsonify({'message': 'User not found'}), 404
    
    data = request.get_json()
    
    if not data:
        return jsonify({'message': 'No data provided'}), 400
    
    # Validate data has at least one health metric
    health_metrics = [
        'weight', 'bmi', 'body_fat', 'muscle_mass', 'water', 
        'visceral_fat', 'bone_mass', 'basal_metabolism', 'protein',
        'calories_consumed', 'calories_burned', 'steps', 'sleep_hours', 'water_intake'
    ]
    
    if not any(metric in data for metric in health_metrics):
        return jsonify({'message': 'At least one health metric is required'}), 400
    
    # Create new health data entry
    new_entry = HealthData(
        user_id=user.id,
        source=data.get('source', 'manual'),
        date=datetime.fromisoformat(data['date'].replace('Z', '+00:00')) if data.get('date') else datetime.utcnow()
    )
    
    # Add all provided metrics
    for metric in health_metrics:
        if metric in data:
            setattr(new_entry, metric, data[metric])
    
    # Save to database
    db.session.add(new_entry)
    db.session.commit()
    
    return jsonify({
        'message': 'Health data added successfully',
        'data': new_entry.to_dict()
    }), 201

@health_data_bp.route('/<int:data_id>', methods=['PUT'])
@jwt_required()
def update_health_data(data_id):
    """Update an existing health data entry"""
    current_user_id = get_jwt_identity()
    user = User.query.filter_by(public_id=current_user_id).first()
    
    if not user:
        return jsonify({'message': 'User not found'}), 404
    
    # Find the health data entry
    entry = HealthData.query.filter_by(id=data_id, user_id=user.id).first()
    
    if not entry:
        return jsonify({'message': 'Health data entry not found'}), 404
    
    data = request.get_json()
    
    if not data:
        return jsonify({'message': 'No data provided'}), 400
    
    # Update fields
    health_metrics = [
        'weight', 'bmi', 'body_fat', 'muscle_mass', 'water', 
        'visceral_fat', 'bone_mass', 'basal_metabolism', 'protein',
        'calories_consumed', 'calories_burned', 'steps', 'sleep_hours', 'water_intake'
    ]
    
    for metric in health_metrics:
        if metric in data:
            setattr(entry, metric, data[metric])
    
    # Update date if provided
    if 'date' in data:
        entry.date = datetime.fromisoformat(data['date'].replace('Z', '+00:00'))
    
    # Save changes
    db.session.commit()
    
    return jsonify({
        'message': 'Health data updated successfully',
        'data': entry.to_dict()
    }), 200

@health_data_bp.route('/<int:data_id>', methods=['DELETE'])
@jwt_required()
def delete_health_data(data_id):
    """Delete a health data entry"""
    current_user_id = get_jwt_identity()
    user = User.query.filter_by(public_id=current_user_id).first()
    
    if not user:
        return jsonify({'message': 'User not found'}), 404
    
    # Find the health data entry
    entry = HealthData.query.filter_by(id=data_id, user_id=user.id).first()
    
    if not entry:
        return jsonify({'message': 'Health data entry not found'}), 404
    
    # Delete entry
    db.session.delete(entry)
    db.session.commit()
    
    return jsonify({
        'message': 'Health data deleted successfully'
    }), 200

@health_data_bp.route('/summary', methods=['GET'])
@jwt_required()
def get_health_summary():
    """Get summary statistics of user health data"""
    current_user_id = get_jwt_identity()
    user = User.query.filter_by(public_id=current_user_id).first()
    
    if not user:
        return jsonify({'message': 'User not found'}), 404
    
    # Get the last 90 days of data
    end_date = datetime.now()
    start_date = end_date - timedelta(days=90)
    
    health_data = HealthData.query.filter_by(user_id=user.id).filter(
        HealthData.date >= start_date,
        HealthData.date <= end_date
    ).order_by(HealthData.date).all()
    
    if not health_data:
        return jsonify({
            'message': 'No health data available',
            'summary': {}
        }), 200
    
    # Convert to DataFrame for easier analysis
    df = pd.DataFrame([data.to_dict() for data in health_data])
    
    # Calculate summary statistics
    summary = {}
    
    # Numeric metrics to analyze
    metrics = [
        'weight', 'bmi', 'body_fat', 'muscle_mass', 'water', 
        'visceral_fat', 'bone_mass', 'basal_metabolism', 'protein',
        'calories_consumed', 'calories_burned', 'steps', 'sleep_hours', 'water_intake'
    ]
    
    for metric in metrics:
        if metric in df.columns and not df[metric].empty and df[metric].notna().any():
            current = df[metric].iloc[-1] if len(df) > 0 else None
            
            metric_summary = {
                'current': float(current) if current is not None else None,
                'min': float(df[metric].min()) if not df[metric].empty else None,
                'max': float(df[metric].max()) if not df[metric].empty else None,
                'avg': float(df[metric].mean()) if not df[metric].empty else None,
                'change': float(df[metric].iloc[-1] - df[metric].iloc[0]) if len(df) > 1 and metric in df.columns else None,
                'change_percent': float((df[metric].iloc[-1] - df[metric].iloc[0]) / df[metric].iloc[0] * 100) if len(df) > 1 and df[metric].iloc[0] != 0 and metric in df.columns else None
            }
            
            summary[metric] = metric_summary
    
    # Add overall stats
    summary['overall'] = {
        'total_records': len(df),
        'first_record_date': df['date'].min() if 'date' in df.columns and not df['date'].empty else None,
        'last_record_date': df['date'].max() if 'date' in df.columns and not df['date'].empty else None
    }
    
    return jsonify({
        'message': 'Health summary generated successfully',
        'summary': summary
    }), 200 

File: backend/routes/healthData.js
const express = require('express');
const router = express.Router();
const { body, validationResult } = require('express-validator');
const auth = require('../middleware/auth');
const HealthData = require('../models/HealthData');

/**
 * @route POST /api/health-data
 * @desc Add new health data entry
 * @access Private
 */
router.post('/', [
  auth,
  body('dataType').notEmpty().withMessage('数据类型是必需的'),
  body('value').isNumeric().withMessage('数值必须是数字'),
  body('timestamp').isISO8601().withMessage('时间戳必须是有效的日期时间格式')
], async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }

  try {
    const { dataType, value, timestamp, note, source = 'manual' } = req.body;
    const userId = req.user.id;
    
    // Check if this exact data point already exists
    const existingData = await HealthData.findOne({
      userId,
      dataType,
      timestamp: new Date(timestamp),
      source
    });
    
    if (existingData) {
      // Update if exists
      existingData.value = value;
      if (note) existingData.metadata.note = note;
      await existingData.save();
      
      return res.json({
        message: '健康数据已更新',
        data: existingData
      });
    }
    
    // Create new health data record
    const healthData = new HealthData({
      userId,
      dataType,
      value,
      timestamp: new Date(timestamp),
      source,
      metadata: note ? { note } : {}
    });
    
    await healthData.save();
    
    res.status(201).json({
      message: '健康数据已保存',
      data: healthData
    });
  } catch (error) {
    console.error('保存健康数据时出错:', error);
    res.status(500).json({ message: '保存健康数据时出错' });
  }
});

/**
 * @route GET /api/health-data
 * @desc Get health data for the user
 * @access Private
 */
router.get('/', auth, async (req, res) => {
  try {
    const userId = req.user.id;
    const { type, startDate, endDate, limit = 100, source } = req.query;
    
    const query = { userId };
    
    if (type) {
      query.dataType = type;
    }
    
    if (source) {
      query.source = source;
    }
    
    if (startDate || endDate) {
      query.timestamp = {};
      
      if (startDate) {
        query.timestamp.$gte = new Date(startDate);
      }
      
      if (endDate) {
        query.timestamp.$lte = new Date(endDate);
      }
    }
    
    const data = await HealthData.find(query)
      .sort({ timestamp: -1 })
      .limit(parseInt(limit));
    
    res.json(data);
  } catch (error) {
    console.error('获取健康数据时出错:', error);
    res.status(500).json({ message: '获取健康数据时出错' });
  }
});

/**
 * @route GET /api/health-data/stats
 * @desc Get statistics for health data
 * @access Private
 */
router.get('/stats', auth, async (req, res) => {
  try {
    const userId = req.user.id;
    const { type, period = '30d' } = req.query;
    
    // Calculate date range based on period
    const endDate = new Date();
    let startDate;
    
    switch (period) {
      case '7d':
        startDate = new Date(endDate.getTime() - 7 * 24 * 60 * 60 * 1000);
        break;
      case '30d':
        startDate = new Date(endDate.getTime() - 30 * 24 * 60 * 60 * 1000);
        break;
      case '90d':
        startDate = new Date(endDate.getTime() - 90 * 24 * 60 * 60 * 1000);
        break;
      case '1y':
        startDate = new Date(endDate.getTime() - 365 * 24 * 60 * 60 * 1000);
        break;
      default:
        startDate = new Date(endDate.getTime() - 30 * 24 * 60 * 60 * 1000);
    }
    
    const query = {
      userId,
      timestamp: {
        $gte: startDate,
        $lte: endDate
      }
    };
    
    if (type) {
      query.dataType = type;
    }
    
    const data = await HealthData.find(query).sort({ timestamp: 1 });
    
    // Calculate statistics
    let stats = {};
    
    if (data.length > 0) {
      if (type) {
        // Single type stats
        const values = data.map(item => item.value);
        stats = {
          count: values.length,
          min: Math.min(...values),
          max: Math.max(...values),
          avg: values.reduce((sum, val) => sum + val, 0) / values.length,
          first: data[0].value,
          last: data[data.length - 1].value,
          change: data[data.length - 1].value - data[0].value,
          changePercent: ((data[data.length - 1].value - data[0].value) / data[0].value) * 100
        };
      } else {
        // Stats for each type
        const typeGroups = data.reduce((groups, item) => {
          if (!groups[item.dataType]) {
            groups[item.dataType] = [];
          }
          groups[item.dataType].push(item);
          return groups;
        }, {});
        
        for (const [type, items] of Object.entries(typeGroups)) {
          if (items.length > 0) {
            const values = items.map(item => item.value);
            stats[type] = {
              count: values.length,
              min: Math.min(...values),
              max: Math.max(...values),
              avg: values.reduce((sum, val) => sum + val, 0) / values.length,
              first: items[0].value,
              last: items[items.length - 1].value,
              change: items[items.length - 1].value - items[0].value,
              changePercent: ((items[items.length - 1].value - items[0].value) / items[0].value) * 100
            };
          }
        }
      }
    }
    
    res.json({
      period,
      startDate,
      endDate,
      stats
    });
  } catch (error) {
    console.error('获取健康数据统计时出错:', error);
    res.status(500).json({ message: '获取健康数据统计时出错' });
  }
});

/**
 * @route DELETE /api/health-data/:id
 * @desc Delete a health data entry
 * @access Private
 */
router.delete('/:id', auth, async (req, res) => {
  try {
    const userId = req.user.id;
    const healthDataId = req.params.id;
    
    const healthData = await HealthData.findById(healthDataId);
    
    if (!healthData) {
      return res.status(404).json({ message: '健康数据记录未找到' });
    }
    
    // Check if this health data belongs to the user
    if (healthData.userId.toString() !== userId) {
      return res.status(403).json({ message: '无权删除此记录' });
    }
    
    await healthData.remove();
    
    res.json({ message: '健康数据记录已删除' });
  } catch (error) {
    console.error('删除健康数据时出错:', error);
    res.status(500).json({ message: '删除健康数据时出错' });
  }
});

module.exports = router; 